qa: fix typehandlerlibrary spotbugs findings

soloturn opened the pull request:
when building with java-11 it ends with:
```
❯ cd subsystems/TypeHandlerLibrary; gradle clean build
...
> Task :subsystems:TypeHandlerLibrary:spotbugsTest
M B PI: The used identifier ?>?2/2??? as field name in class org.terasology.persistence.typeHandling.InMemorySerializerTest$TypeGetter.STRING in source file ?>?3/2??? shadows the publicly available identifier from the Java Standard Library.  In InMemorySerializerTest.java
M C EC: Using .equals to compare two byte[]'s, (equivalent to ==) in org.terasology.persistence.typeHandling.coreTypes.factories.BytesTypeHandlerTest.byteArraySerializeDeserialize()  At BytesTypeHandlerTest.java:[line 42]
M C EC: Using .equals to compare two byte[]'s, (equivalent to ==) in org.terasology.persistence.typeHandling.coreTypes.factories.BytesTypeHandlerTest.byteArraySerializeDeserialize()  At BytesTypeHandlerTest.java:[line 45]
M C EC: Using .equals to compare two byte[]'s, (equivalent to ==) in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeBytes()  At InMemorySerializerTest.java:[line 309]
M C EC: Using .equals to compare two byte[]'s, (equivalent to ==) in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeByteBuffer()  At InMemorySerializerTest.java:[line 338]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeString()  At InMemorySerializerTest.java:[line 82]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeStrings()  At InMemorySerializerTest.java:[line 113]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.template(PersistedData)  At InMemorySerializerTest.java:[line 136]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeBoolean()  At InMemorySerializerTest.java:[line 255]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeBytes()  At InMemorySerializerTest.java:[line 319]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeByteBuffer()  At InMemorySerializerTest.java:[line 348]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.serializeNull()  At InMemorySerializerTest.java:[line 392]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.checkIsArray(PersistedData)  At InMemorySerializerTest.java:[line 434]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.checkIsNumber(PersistedData)  At InMemorySerializerTest.java:[line 458]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.checkValueArray(PersistedData, PersistedData, Set)  At InMemorySerializerTest.java:[line 486]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.lambda$checkValueArray$8(Executable)  At InMemorySerializerTest.java:[line 512]
H C RV: org.junit.jupiter.api.Assertions.assertThrows(Class, Executable) not thrown in org.terasology.persistence.typeHandling.InMemorySerializerTest.lambda$checkValueArray$4(Executable)  At InMemorySerializerTest.java:[line 497]
M P UuF: Unused field: org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactoryTest$SomeClass.t  In ObjectFieldMapTypeHandlerFactoryTest.java
M P SIC: Should org.terasology.persistence.typeHandling.FutureTypeHandlerTest$ResultCaptor be a _static_ inner class?  At FutureTypeHandlerTest.java:[lines 41-50]
M P UuF: Unused field: org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactoryTest$SomeClass.list  In ObjectFieldMapTypeHandlerFactoryTest.java
SpotBugs ended with exit code 1

> Task :subsystems:TypeHandlerLibrary:spotbugsMain
SpotBugs ended with exit code 1
```


see also #3859

soloturn committed:
typehandlerlibrary qa

as a workaround, add dependency for spotbugs annotations, and create a
[ticket there so the spotbugs gradle plugin would include the dependency]
(https://github.com/spotbugs/spotbugs-gradle-plugin/issues/1018).

see here a motivation for inner type last rule. if we leave the rule we
should respect it, otherwise disable the rule.
https://stackoverflow.com/questions/25158939/what-motivation-is-behind-checkstyle-inner-type-last-rule

especially in tests just quite the warnings, in most cases the code is
deliberate as it is.

see #3859 and see bugs:
* https://github.com/pmd/pmd/issues/4731
* https://github.com/spotbugs/spotbugs-gradle-plugin/issues/1018
* https://github.com/spotbugs/spotbugs/issues/2667

soloturn committed:
chore: kotlin jps updated itself

jdrueckert commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/serializers/Serializer.java:
```diff
@@ -22,6 +22,9 @@
  * methods that use {@link #serializeToPersisted(Object, TypeInfo)} and {@link #deserializeFromPersisted(PersistedData,
  * TypeInfo)}.
  */
+// log statements after an if are marked as false positive, suppress.
+// see pmd bug: https://github.com/pmd/pmd/issues/4731
+@SuppressWarnings("PMD.GuardLogStatementJavaUtil")
```
is this about "[Logger calls should be surrounded by log level guards.](https://pmd.github.io/pmd-6.55.0/pmd_rules_java_bestpractices.html#guardlogstatement)" PMD findings?
I don't think they're false positives per se, as generally it can be good practise to have these guards.
However, in our case, I don't think the rule actually provides much benefit. we can discuss this in the reviver round, but I'd say that rule is a good candidate for removal instead of suppressing it everywhere.

soloturn replied:
you were right. there is an alternative: https://stackoverflow.com/questions/35096746/substituting-parameters-in-log-message-and-add-a-throwable-in-log4j-2, implemented. upgraded pmd as well to 7.0.0-rc4 to make sure all keeps working. there is one bug though, and i updated: https://github.com/pmd/pmd/issues/4731.

i removed now all the guardlog suppresses, and created pull request to remove the rule: https://github.com/MovingBlocks/TeraConfig/pull/23

can it be that this already was decided? a comment in the teraconfig pmd rules already tried to exclude it.

jdrueckert commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/CollectionTypeHandlerFactory.java:
```diff
@@ -24,7 +24,7 @@
 public class CollectionTypeHandlerFactory implements TypeHandlerFactory {
     private static final Logger LOGGER = LoggerFactory.getLogger(CollectionTypeHandlerFactory.class);
 
-    private ConstructorLibrary constructorLibrary;
+    @SuppressWarnings("PMD.UnusedLocalVariable") private ConstructorLibrary constructorLibrary;
```
why are we suppressing this here?

soloturn replied:
code is here sincd 2018 and neither intellij or pmd are able to judge if it is good or not.

jdrueckert commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/inMemory/PersistedBytes.java:
```diff
@@ -5,6 +5,7 @@
 
 import java.nio.ByteBuffer;
 
+@SuppressWarnings({"PMD.ArrayIsStoredDirectly", "PMD.MethodReturnsInternalArray"})
```
why are we suppressing these here?

soloturn replied:
code was written 2008, does it really make sense to rewrite it, and clone the bytes? if something is working for so long imo it is better to focus on newly introduced vulnerabilites.

jdrueckert replied:
The age of code does not render it unworthy of refactoring - on the contrary: legacy code is more likely to need refactoring as it might be using outdated language features or programming patterns. The fact that it works for so long is irrelevant, as we're not talking about fixing it, but about refactoring it to improve its quality.
Also, considering that we have performance issues for a long time, I don't expect them to originate mainly from newly introduced code.

We can configure qualitative scans in a way that only newly introduced flaws block the pipeline, so there's no urgent need to suppress potentially valid complaints by our scan tools. We should rather address them in dedicated efforts to improve our code (especially the ages-old areas).

If the code's age is the main or even only reason for suppressing it, I don't think we should suppress.

soloturn replied:
the goal is to get rid of warnings, if there are 100 warnings, and one is added, it is difficult to spot, and it will become 101. if the old ones are suppressed there is 0 warnings, and it will stay 0.

having this goal leaves 2 options: fix or accept (i.e. works as designed). for the alternative, fix, what would be your suggestion?

jdrueckert replied:
I'd argue that the goal should not be to get rid of warnings (except maybe to reduce noise) but to improve the quality of our code. If we do that properly, then the warnings will be reduced.

As mentioned, scan tooling can be configured to only block the pipeline on issues newly introduced in a PR. We already have that, for instance, for Checkstyle, see for instance [here](https://github.com/MovingBlocks/Terasology/runs/18439649448):
![image](https://github.com/MovingBlocks/Terasology/assets/29981695/c28f8760-685b-422d-8ab0-725bd71895f2)

In the checkstyle integration for Jenkins, you can then filter for new findings:
![image](https://github.com/MovingBlocks/Terasology/assets/29981695/4dec5215-6198-4f0f-9eb5-7d82bc83dce4)
![image](https://github.com/MovingBlocks/Terasology/assets/29981695/6a6ee116-bb19-4ca7-bbad-a72378b89d02)

So again, the goal should not be to get rid of warnings for the sake of getting rid of warnings.

---

From the top of my head right now, I don't know how to fix it. But neither means that we cannot fix it nor that we shouldn't attempt to fix it.
For fixing it, we would need to have a dedicated closer look at the code and the PMD rule to understand what the motivation of the rule is (aka why the code that results in the finding is bad) and think about how to improve it.

soloturn replied:
if you really want to fix it or get it fixed, TODO is the better concept imo. should i add one? other possiblity would be to create a ticket where one can describe findings up to now. 

i have no findings up to now apart from: the code is very old, side effects are possible, maybe the code depends on side effects. one needs to keep in mind why such a pmd rule exists. it is a security warning. if it is some API, the caller is untrusted, it can change the values from outside. in terasology case it is a game, copying stuff slows it down, and you have nobody calling from outside.

soloturn replied:
see here as well:
https://www.appsloveworld.com/java/100/140/reasoning-behind-arrayisstoreddirectly-rule-of-pmd

jdrueckert replied:
Whether or not TODOs are better here highly depends. TODOs pose the danger of losing information by decoupling the underlying issue (description) from the perceived (subjective) action to mitigate it.

I had a quick look at the `PMD.ArrayIsStoredDirectly`and `PMD.MethodReturnsInternalArray` rules and the surrounding context. I agree that the code was likely written the way it was because the main focus was on performance. Personally, coming from a security point of view, I'm hesitant to just accept this as "likely not a problem" without a proper criticality assessment of this security flaw. However, at this point in time I have neither the energy nor the time to do this assessment the way I think it should be done.

So... how can we go from here. I would propose two options and you're welcome to pursue either of them:
A) keep the warnings and add a TODO comment to do a criticality assessment to decide whether they should be suppressed or remediated.
B) suppress the warnings and "replace" them with a TODO comment referencing an issue that outlines the reasoning behind the rules and discusses possible mitigating steps including the performance-vs-security trade-off.

soloturn replied:
took approach B).

jdrueckert replied:
my bad, should've been more explicit... the TODO comment should reference a **GitHub** issue.
> B) suppress the warnings and "replace" them with a TODO comment referencing ~~an issue~~ **edit: a GitHub issue** that outlines the reasoning behind the rules and discusses possible mitigating steps including the performance-vs-security trade-off.

jdrueckert commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/inMemory/arrays/PersistedBooleanArray.java:
```diff
@@ -9,6 +9,7 @@
 import java.util.Arrays;
 import java.util.Iterator;
 
+@SuppressWarnings({"PMD.ArrayIsStoredDirectly", "PMD.MethodReturnsInternalArray"})
```
why are we suppressing these here?

soloturn replied:
code was written 2008, does it really make sense to rewrite it, and clone the bytes? if something is working for so long imo it is better to focus on newly introduced vulnerabilites.

jdrueckert replied:
see above

jdrueckert replied:
same here

jdrueckert commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/FutureTypeHandlerTest.java:
```diff
@@ -77,4 +53,29 @@ public void testRecursiveType() {
+        }
+    }
+
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings
```
why is this necessary?

soloturn replied:
added a comment. message that one should consider making the class static. but the test has it non-static.

jdrueckert replied:
> but the test has it non-static.

Sorry, you'll have to elaborate that a bit more, I don't follow.

soloturn replied:
updated the comment @jdrueckert .

jdrueckert replied:
Based on the reasoning in my comment above (line 68), I'd propose to adjust this as follows and remove the comment you added (lines 68-70):
```suggestion
    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings(value = "SIC_INNER_SHOULD_BE_STATIC", justification = "Test code is not performance-relevant, flagged inner ResultCaptor class is a mock with dynamic behavior and cannot be static.")
```

`value` filters the exact spotbugs finding that should be suppressed, which avoids suppressing other (future) findings that might be valid
`justification` documents why we can ignore this finding here in a concise way and as part of the suppression annotation

jdrueckert commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/InMemorySerializerTest.java:
```diff
@@ -38,8 +38,11 @@
 import java.util.stream.Stream;
 
 
+// spotbugs does thinks Assertions.assertThrows does not throw the exception, even if
+// test is successful. see bug: https://github.com/spotbugs/spotbugs/issues/2667.
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings
```
seems to be fixed with https://github.com/spotbugs/spotbugs/pull/2648 but not released yet
we don't fail for spotbugs findings at the moment, so I would not suppress this now but wait for the spotbugs patch

soloturn replied:
agree and took it out.

soloturn replied:
new spotbug version now included, warnings are gone.

jdrueckert commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactoryTest.java:
```diff
@@ -20,6 +20,10 @@
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 
+// both, pmd and spotbugs complain about not used private fields, suppress in
+// the static test class, but fields are checked. suppress.
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings
+@SuppressWarnings("PMD.UnusedPrivateField")
```
I'm not entirely sure we want to suppress this as it might be a valid indicator of improvement potential. Either the private fields are indeed unused then this begs the question why they're in here in the first place. And if they are not explicitly used but implicitly required for the tests, then I would argue that this asks for refactoring due to nontransparent side effects.

soloturn replied:
it is "just" a type handler test and checks the type? to test nothing needs to be implemented using the fields.

jdrueckert replied:
> to test nothing needs to be implemented using the fields.

Sorry, you'll have to elaborate on this, I don't follow.

soloturn replied:
moved the suppressions to the class which really causes it, when i tried that initially it failed, but cannot reproduce how. its a minimal test case, no point in using these private fields other than by typehandler.

soloturn committed:
typehandlerlibrary qa, jdrueckert feedback

especially take out pmd guardlog suppresses. fix part of them
in a catch even if not necessary from a performance point of
view. and create a pull request towards teraconfig to not
use the guardlog rule any more:
https://github.com/MovingBlocks/TeraConfig/pull/23

Co-authored-by: jdrueckert <jd.rueckert@googlemail.com>

soloturn commented:
@jdrueckert anything else missing?

soloturn commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactoryTest.java:
```diff
@@ -48,4 +43,12 @@ public void testObject() {
+
+    // the minimal test case leads to a warning of unused private fields, in both spotbugs and pmd, suppress.
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings
+    @SuppressWarnings("PMD.UnusedPrivateField")
```
moved suppression to inner class now.

jdrueckert commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/FutureTypeHandlerTest.java:
```diff
@@ -77,4 +53,32 @@ public void testRecursiveType() {
+        }
+    }
+
+    // the test verfies that bot, static class (above), and class work. with this typehandler.
```
Spotbugs complains about the inner class `ResultCaptor` not being static. I'm not convinced, that the statement "the test verfies that bot, static class (above), and class work. with this typehandler." is correct. The test verifies that for recursive types, the expected typehandler (`FutureTypeHandler`) is fetched. For this, it uses `ResultCaptor` as a mock to first capture the result and later assert on it.

The reason why it doesn't work making the inner `ResultCaptor` class static is that this mock is not a static mock - it doesn't return a pre-defined value but executes the "real" method and captures its result.

Furthermore, the reason why Spotbugs complains about this is that the inner class not being static may result in performance issues due to increased instance sizes and object lifetimes. However, as this is a test, we don't care as much about performance.

jdrueckert commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/factories/ObjectFieldMapTypeHandlerFactoryTest.java:
```diff
@@ -48,4 +43,12 @@ public void testObject() {
     }
+
+    // the minimal test case leads to a warning of unused private fields, in both spotbugs and pmd, suppress.
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings
```
```suggestion
    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings(value = "UUF_UNUSED_FIELD", justification = "Direct member access is not expected. TypeHandler factory dynamically loads type handlers on type handler creation based on member types of input class TypeInfo. ")
```

soloturn committed:
update spotbugs version.

now AssertThrows error is fixed:
https://github.com/spotbugs/spotbugs/issues/2667

jdrueckert commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/factories/CollectionTypeHandlerFactory.java:
```diff
@@ -24,7 +24,8 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(CollectionTypeHandlerFactory.class);
 
-    private ConstructorLibrary constructorLibrary;
+    // PMD and intellij cannot judge if this is good or not. suppress.
+    @SuppressWarnings("PMD.UnusedLocalVariable") private ConstructorLibrary constructorLibrary;
```
I'm still questioning whether we want to suppress this. Again, the age of the code is not a valid reason IMO.

jdrueckert replied:
I believe, we can remove this field from the `CollectionTypeHandlerFactory` class.
I tested it locally and didn't see any issues. My assumption is that this is a left-over from refactoring `ConstructorLibrary#getCollectionCopyConstructor` to a static method. This allowed the call in `CollectionTypeHandlerFactory#create` to use the method without instantiating `ConstructorLibrary`.

jdrueckert committed:
Merge branch 'develop' into typehandlerlibrary-qa
